[FEATURE] Afterlife temperaments

scribblecrumb opened the pull request:
<!-- Write BELOW The Headers and ABOVE The comments else it may not be viewable. -->
<!-- You can view CONTRIBUTING.md for a detailed description of the pull request process. -->
<!-- Be sure to name your PR something descriptive and succinct; include Bugfix: Feature: Enhancement: or Content: in the title to describe what type of PR it is. -->
<!-- IF YOU ARE DOING A BUGFIX: Please target the latest release branch if the bug that you are fixing is also present in the latest release. -->

## About The Pull Request
This adds afterlife temperament values.  These currently don't affect the game in any way, we're just getting this stuff in piecemeal so that PRs are manageable.

This adds an `Afterlife` class which is used for the new `game.starclan` and `game.dark_forest` attributes.  This class tracks facets and temperament.  By default, all facets are set to 8, meaning that an empty afterlife will have an average facet default and stoic temper.

In order to avoid handling an insane number of dead cats and to ensure that the "important" dead ones are the ones with influence, we only allow leaders, deputies, medicine cats, and guides to affect facets.  

Rather than fully re-calcing facets on each timeskip, we keep track of cats who have changed afterlife via `game.updated_afterlife_cats`.  When we timeskip, each cat is checked that they're in:
- A new afterlife (to avoid weird interactions when a cat is flip flopped a bunch between afterlives)
- An appropriate rank (or the guide)
Then the cat is added to the `afterlife.influencing_cats` set and we do some math to add each of their facets to the afterlife's facet averages.  If the cat was removed from an afterlife, then they're also removed from it's `influencing_cats` and facet average

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage senior developers from merging your PR! -->

## Why This Is Good For ClanGen
Continues on our journey to making affinities do things!  Once we know afterlife tempers, we can start determining what each afterlife thinks about each cat, and how that affects their affinity to that afterlife.
<!-- If this is a bug fix, you can remove this section. -->
<!-- Please add a short description of why you think these changes would benefit the game. -->
<!-- If this PR was the result of discussion/testing on the discord, please add a link to the discord conversation here. -->
<!-- If you have multiple features that can stand on their own, or unrelated bugfixes, please create separate PRs for them. -->

## Proof of Testing
I'll use the debugger to step through this, since nothing displays these tempers in-game yet. StarClan is started with a size of 5.

This was also done without the rank check, as that woulda been an annoying thing to account for when checking that the basic adding/removing statements are working.  However, I did test afterwards with the rank check and confirmed that only leaders, meddies, and deputies could make it through and change the facets.  I just wanted this debugging to display that I'd accounted for many different situations.

Cat exiled and sent to UR
<img width="691" height="209" alt="image" src="https://github.com/user-attachments/assets/21cce2ce-f58f-4da0-8bd8-cf10f0ff25d8" />

<img width="372" height="265" alt="image" src="https://github.com/user-attachments/assets/a2722766-cd0b-4579-8a4a-5fa55671b5c3" />
<img width="254" height="221" alt="image" src="https://github.com/user-attachments/assets/1ecf7fb5-c01b-4cff-9a38-2b0e23430c9e" />

SC and DF are unchanged (and you can see that DF is defaulting to average 8 values, since it's currently empty)

UR cat sent to SC
<img width="519" height="243" alt="image" src="https://github.com/user-attachments/assets/95c07926-3c44-4ffc-b906-c9343918d79a" />

<img width="317" height="244" alt="image" src="https://github.com/user-attachments/assets/c153f50a-0435-4335-b502-860be6a4cab4" />
<img width="264" height="218" alt="image" src="https://github.com/user-attachments/assets/4a495ecc-6ed8-44cc-87c1-77a10e0d5e5d" />

SC changed, DF did not

SC cat sent to DF
<img width="488" height="227" alt="image" src="https://github.com/user-attachments/assets/9c46af5d-5319-4e64-9e39-8374eb77326f" />

<img width="274" height="232" alt="image" src="https://github.com/user-attachments/assets/a511ef1b-5d16-42f3-9641-51c48f0c8c8e" />
<img width="283" height="247" alt="image" src="https://github.com/user-attachments/assets/44bfe695-cf52-42f9-b028-0d9c1ac969c1" />

SC and DF changed, SC lost the cat and DF gained it

Cat killed and sent to SC
<img width="432" height="239" alt="image" src="https://github.com/user-attachments/assets/f7bdd794-0d3a-4f42-a311-3419fb93dd3f" />

<img width="271" height="226" alt="image" src="https://github.com/user-attachments/assets/9657d189-0694-4528-b664-197a7ac986f8" />
<img width="283" height="226" alt="image" src="https://github.com/user-attachments/assets/673bbc87-8231-444c-b334-590b4f3d1d65" />

SC changed, DF did not

Cat killed and sent to DF (after initially being sent to SC)
<img width="637" height="365" alt="image" src="https://github.com/user-attachments/assets/ed48bf5b-5794-44da-812f-b03f68c66d8a" />

<img width="295" height="248" alt="image" src="https://github.com/user-attachments/assets/4a43f852-e656-4693-8d2d-12dc72bdb918" />
<img width="279" height="218" alt="image" src="https://github.com/user-attachments/assets/e6e37686-552a-4578-b6fd-87f44379e80a" />

DF changed, SC did not

Set of cats to update facets with is reset after all updates are applied
<img width="324" height="105" alt="image" src="https://github.com/user-attachments/assets/439a6dbd-8ca7-460e-85a6-56dae813a3a8" />


Now with the rank check on:
dead deputy
<img width="483" height="186" alt="image" src="https://github.com/user-attachments/assets/ed73b5d7-a1cd-4e69-9de0-1fa0431b2cc3" />
<img width="271" height="242" alt="image" src="https://github.com/user-attachments/assets/6acd3721-6d24-432d-9123-8d40146c7592" />

Changed SC as intended

Dead warrior
<img width="434" height="169" alt="image" src="https://github.com/user-attachments/assets/3c0960e8-4c4a-414e-a8ce-dd50e8a31a27" />
<img width="476" height="242" alt="image" src="https://github.com/user-attachments/assets/34e4f260-46d2-434f-acca-4f3f1643aef9" />

Gets continued instead of adjusting facets

<!-- Include any screenshots, debugging steps, or links to beta testing threads here. At least one form of proof of testing is REQUIRED for all new content. You must be able to run the code locally before you PR it here. -->

## Changelog/Credits
- Adds temperaments & facets to StarClan and the Dark Forest based on the high ranking cats within each (guide, leaders, deputies, medicine cats).  Currently has no affect on gameplay and is not viewable in-game.
<!-- Include any changes that should be made to the changelog of the game here or any changes to the credits file of the game. -->
<!-- This is just for easy access later for senior developers gathering this information; this process is not automated. -->

scribblecrumb committed:
adding afterlife classes and calc-ing temper

scribblecrumb committed:
various fixes

scribblecrumb committed:
allow instructors

scribblecrumb committed:
fix afterlives in tests

scribblecrumb committed:
remove unneeded statement

silly scribble, this guy already gets added during cat load

scribblecrumb committed:
test fix

larkgz commented on scripts/clan.py:
```diff
@@ -1309,41 +1295,117 @@ def __repr__(self):
+        else:
+            old_size = self.size - 1
+
+        return int(((old_avg * old_size) + new_value) / self.size)
```
Converting to int here will result in quickly accumulating roundoff errors, and may also give different averages for the same numbers depending on the order that they're added.

larkgz commented on scripts/clan.py:
```diff
@@ -1309,41 +1295,117 @@ def __repr__(self):
+            old_size = self.size + 1
+            new_value = -new_value
+        else:
+            old_size = self.size - 1
```
Can you pass in `old_size` as a parameter instead of recalculating it here?

larkgz commented on scripts/clan.py:
```diff
@@ -1309,41 +1295,117 @@ def __repr__(self):
+                self.new_average(_int, getattr(cat.personality, name), removal=removal),
+            )
+
+    def new_average(self, old_avg: int, new_value: int, removal: bool) -> int:
```
* Please add a docstring to this function.
* If it's only going to be used as a helper in this class, please add an underscore to the name.
* I think `new_average` is a bit unclear. Maybe something like `_add_to_average` or something?

larkgz commented on scripts/clan.py:
```diff
@@ -1309,41 +1295,117 @@ def __repr__(self):
+                self,
+                name,
+                self.new_average(_int, getattr(cat.personality, name), removal=removal),
+            )
```
This is kind of hard to understand at a glance. 

Since it's all fixed attributes, I don't think that using setattr, getattr, and zip is strictly necessary here. There'll be some repetition, but I think that's a fair exchange for better readability.

larkgz commented on scripts/clan.py:
```diff
@@ -1309,41 +1295,117 @@ def __repr__(self):
+    def temperament(self) -> str:
+        return get_temper_alignment(self.sociability, self.aggression)
+
+    def change_facets(self, cat: Cat, removal: bool = False):
```
* Please add a docstring to this function.
* Might want to rename `removal` to `is_removal` so it's more quickly understandable as a boolean?
* Maybe rename to `change_facets_by_cat`? `change_facets` sort of feels like it's setting all the facets to a given value.

larkgz commented on scripts/events.py:
```diff
@@ -99,6 +98,43 @@ def one_moon(self):
+                continue
+            if (
+                not c.status.rank
+                in (
```
```suggestion
                c.status.rank
                not in (
```
Same thing but a bit easier to read.

larkgz commented on scripts/cat/status.py:
```diff
@@ -357,6 +357,13 @@ def is_other_clancat(self) -> bool:
         return self.rank == CatRank.LEADER
 
+    @property
+    def newly_joined_group(self) -> bool:
```
`joined_group_this_moon` might be clearer?

larkgz commented on scripts/events.py:
```diff
@@ -99,6 +98,43 @@ def one_moon(self):
+            ):
+                continue
+
+            # first change facets of the group they joined
+            if c.status.group == CatGroup.STARCLAN:
+                game.starclan.change_facets(c)
+            elif c.status.group == CatGroup.DARK_FOREST:
+                game.dark_forest.change_facets(c)
+
+            if len(c.status.group_history) == 1:
+                continue
+
+            if c.status.group_history[-2]["moons_as"] == 0:
+                # this cat never had their facets added to this group, so we won't try to remove them
+                continue
+
+            prior_group = c.status.group_history[-2]["group"]
+
+            if prior_group == CatGroup.STARCLAN_ID:
+                game.starclan.change_facets(c, removal=True)
+            elif prior_group == CatGroup.DARK_FOREST_ID:
+                game.dark_forest.change_facets(c, removal=True)
+        game.updated_afterlife_cats.clear()
```
Not fully sold on this approach (i.e. looping over just updated_afterlife_cats and then checking history to correct any removals). It feels a bit like premature optimization that's more difficult to maintain. 

We already iterate over every cat, so imo it's fine to just recalculate the afterlife facets from everyone each moon.

scribblecrumb replied:
Doing that math every moon regardless of whether or not the afterlife population changed seems wasteful though.  And players who turn off fading inevitably end up with large afterlives (even if fading is on, they can get pretty big), which would mean a lot of extraneous calculation occurring.

I do agree that the history checks are a little annoying to deal with. A different approach could be storing a list (actually a set would prolly be best) of IDs for each cat who influences the facets. Then we use the len for the average calc and we just check for the cat ID's presence when checking if their influence needs to be removed from the group.  That would be more maintainable in the long run, and easier to understand at a glance.

larkgz reviewed (changes requested):
Please read the last comment first, since it's the most important.

scribblecrumb committed:
docstrings and attribute renames

scribblecrumb committed:
use floor division instead of division

avoids float rounding issues

scribblecrumb committed:
readability pass

scribblecrumb committed:
switch to a set of cat IDs

now we gather a set of IDs for each cat who influences the afterlife and use it's len for the average calcs. this makes it easier to tell if a cat was already included or not.

scribblecrumb committed:
fix accidental rename

scribblecrumb committed:
fix guide influence

scribblecrumb committed:
explanatory comment

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into afterlife-temper

larkgz commented on scripts/clan.py:
```diff
@@ -1309,41 +1307,149 @@ def __repr__(self):
+        if is_removal:
+            new_value = -new_value
+
+        return ((old_avg * old_size) + new_value) // new_size
```
Unfortunately, this is still going to result in accumulated roundoff errors. The issue is that it's using the old average, which is already based off a rounded value, and then rounding the value again.

It might make sense to keep the totals (e.g. `self._total_aggression`) instead of retrieving them by multiplying `old_avg * old_size`.

larkgz commented on scripts/clan.py:
```diff
@@ -1309,41 +1307,149 @@ def __repr__(self):
+class Afterlife:
     """
-    TODO: DOCS
+    Currently just used for tracking temperament & facets. All facets default to 8 if dead_cats is empty.
```
```suggestion
    Currently just used for tracking temperament & facets. All facets default to 8 if influencing_cats is empty.
```

scribblecrumb committed:
avoiding roundoff erros

scribblecrumb committed:
doc fix

larkgz commented on scripts/clan.py:
```diff
@@ -491,10 +492,22 @@ def load_clan(self):
             )
 
+        # can't put this in post initialization bc guide isn't made before that func
+        self.load_guide_influence()
```
For some reason, it seems like the guide is being loaded into the afterlife influence another time, prior to this function being called. I have no idea where.

I think either:
* Remove this function
* Check when adjusting facets by cat if the cat is already an influencer or not. I guess one way of doing this would be checking if the length of the set is the same after trying to add/remove a cat.

scribblecrumb replied:
Oh! Theory: I bet this is happening when the guide is a leader/dep/medicine cat.  Since those are the qualifiers for influencing temperament, they get included during the initial cat loading process.  Then when the guide is loaded, they get added to the temper *again*.  This isn't an issue for guides who are any other rank.  

So I think I will go ahead and check if the cat is already an influencer before adjustments are made, as this function is still necessary for those non-leadership role guides.

larkgz commented on scripts/clan.py:
```diff
@@ -1309,41 +1307,150 @@ def __repr__(self):
+        self._aggress: int = 0
+        self._stable: int = 0
+
+        self.total_aggression: int = 0
+        self.total_lawfulness: int = 0
+        self.total_sociability: int = 0
+        self.total_stability: int = 0
```
Can you rename these variables with an underscore in front?

larkgz commented on scripts/clan.py:
```diff
@@ -1309,41 +1307,150 @@ def __repr__(self):
+        if not self.influencing_cats:
+            return 8
+        else:
+            return self._aggress
+
+    @aggression.setter
+    def aggression(self, val):
+        self._aggress = val
+
```
Now that I think about it, I'm not sure if this class even needs public setters for law, sociability, aggression, and lawfulness, since setting these values from outside will throw off the average calculations.

Possibly print an error message instead?

scribblecrumb committed:
FIX: feedback changes

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into afterlife-temper

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into afterlife-temper

archanyhm commented on scripts/clan.py:
```diff
@@ -1309,41 +1307,160 @@ def __repr__(self):
-        TODO: DOCS
+        Adjusts the afterlife's facet averages according to the facets of the given cat
+        :param cat: The cat object adjust facets by
+        :param is_removal: Set True if the cat's facets are being removed from the afterlife's
```
Perhaps rename `is_removal` to `do_remove`?

archanyhm commented on scripts/cat/personality.py:
```diff
@@ -4,12 +4,14 @@
 
-    facet_types = ["lawfulness", "sociability", "aggression", "stability"]
-    facet_range = [0, 16]
+    facet_types = constants.facet_types
+    facet_range = constants.facet_range
```
Not sure what the point is to move them into the `constants` file. Are they used anywhere else?

scribblecrumb replied:
At the moment, no (though I think there was an iteration of this PR where I was using this it in the Afterlife class, which likely prompted this change), but we've been moving this sort of information to `constants` lately. I personally like having it consolidated in one file, but I understand if we'd rather not in this case.

archanyhm replied:
I have to think about this some more. Given that this is relatively easy to refactor down the line, I'll mark this as resolved and make a note to bring it up during the code-style-guidelines-topic that we were talking about recently.

archanyhm commented on scripts/cat/status.py:
```diff
@@ -357,6 +357,13 @@ def is_other_clancat(self) -> bool:
         return self.rank == CatRank.LEADER
 
+    @property
+    def joined_group_this_moon(self) -> bool:
```
Perhaps rename to `did_join_group_this_moon` instead.
(More consistent with `is_`, `has_`, `do_`-nomenclature for boolean-members)

archanyhm commented on scripts/clan.py:
```diff
@@ -1309,41 +1307,160 @@ def __repr__(self):
+    def aggression(self, value):
+        print(
+            "ERROR: Afterlife aggression cannot be set manually as it is meant to be calculated from the currently dead cats."
+        )
```
Raise exception instead.

archanyhm commented on scripts/clan.py:
```diff
@@ -1309,41 +1307,160 @@ def __repr__(self):
+    def sociability(self, value):
+        print(
+            "ERROR: Afterlife sociability cannot be set manually as it is meant to be calculated from the currently dead cats."
+        )
```
Raise exception instead

archanyhm commented on scripts/clan.py:
```diff
@@ -1309,41 +1307,160 @@ def __repr__(self):
+    def lawfulness(self, value):
+        print(
+            "ERROR: Afterlife lawfulness cannot be set manually as it is meant to be calculated from the currently dead cats."
+        )
```
Raise exception instead

archanyhm commented on scripts/clan.py:
```diff
@@ -1309,41 +1307,160 @@ def __repr__(self):
+    def stability(self, value):
+        print(
+            "ERROR: Afterlife aggresstabilitysion cannot be set manually as it is meant to be calculated from the currently dead cats."
+        )
```
Raise exception instead

archanyhm commented on scripts/events.py:
```diff
@@ -100,6 +99,45 @@ def one_moon(self):
+                if c.ID in game.starclan.influencing_cats:
+                    game.starclan.adjust_facets_by_cat(c, is_removal=True)
+
+        game.updated_afterlife_cats.clear()
```
Pull into smaller helper function(s) since `one_moon` is already quite hefty.

archanyhm commented on scripts/clan.py:
```diff
@@ -1309,41 +1307,160 @@ def __repr__(self):
+        :param total: The facet's total value derived from all influencing cats
+        :param num_of_influencers: The number of cats influencing the average
+        :return: The adjusted average
         """
-        white = pygame.Surface((sprites.size, sprites.size))
-        fade_level = 0
-        if cat.dead:
-            for f in self.forgotten_stages:  # pylint: disable=consider-using-dict-items
-                if cat.dead_for in range(
-                    self.forgotten_stages[f][0], self.forgotten_stages[f][1]
-                ):
-                    fade_level = f
-        white.fill((255, 255, 255, fade_level))
-        return white
+        return total // num_of_influencers
```
Probably rename to something that expresses more clearly that this relates to facets.

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into afterlife-temper

scribblecrumb committed:
address feedback

scribblecrumb committed:
formatter

archanyhm commented on scripts/clan.py:
```diff
@@ -491,10 +492,22 @@ def load_clan(self):
+    def load_guide_influence():
+        """
+        Adds guide's facet influences to their current afterlife
+        """
```
Potential nitpick, but does `load_guide_influence` really describe best what the function does? Perhaps `add_guide_facet_influence` or `apply_guide_facet_influence` would be more in-line?
(This is not a change-request; just an idea)

archanyhm replied:
(So if you disagree, feel free to just resolve this)

archanyhm reviewed (approved):
Not sure if @larkgz has anything else to say, but lgtm.

scribblecrumb committed:
rename func
